src/preamble.js

```diff
@@ -130,14 +130,23 @@ function ccall(ident, returnType, argTypes, args, opts) {
     });
   }
 #else // EMTERPRETIFY_ASYNC
+  if (typeof Asyncify === 'object' && Asyncify.currData !== null) {
 #if ASSERTIONS
-  assert(!(opts && opts.async), 'async call is only supported with Emterpretify for now, see #9029');
+    assert(opts && opts.async, 'The call to ' + ident + ' is running asynchronously. If this was intended, add the async option to the ccall/cwrap call.');
+    assert(Asyncify.asyncFinalizers.length === 0, 'Cannot have multiple async ccalls in flight at once');
 #endif
+    return new Promise(function(resolve) {
+      Asyncify.asyncFinalizers.push(function(ret) {
+        if (stack !== 0) stackRestore(stack);
+        resolve(convertReturnValue(ret));
+      });
+    });
+  }
 #endif // EMTERPRETIFY_ASYNC
 
   ret = convertReturnValue(ret);
   if (stack !== 0) stackRestore(stack);
-#if EMTERPRETIFY_ASYNC
+#if EMTERPRETIFY_ASYNC || ASYNCIFY
```

Comment:
We can directly return a resolved Promise if run in `async` mode, but no `Asyncify.handleSleep` ever occurred.